src/app/interaction-model/InvokeResponder.h

```diff
@@ -16,19 +16,44 @@
  */
 #pragma once
 
+#include <app/MessageDef/StatusIB.h>
 #include <app/data-model/WrappedStructEncoder.h>
 #include <lib/core/CHIPError.h>
 
 namespace chip {
 namespace app {
 namespace InteractionModel {
 
-/// Aids in writing invoke replies:
-///   - Has access to the raw data encoder for response data content (via `ResponseEncoder`)
-///   - `Complete` handling:
-///      - MUST be called exactly once
-///      - called with an error, makes TLV data written to be discarded (Invokes may only either
-///        succeed or fully fail)
+/// Handles encoding of an invoke response for a specific invoke request.
+///
+/// This class handles a single response (i.e. a CommandDataIB within the
+/// matter protocol) and is responsible for constructing its corresponding
+/// response (i.e. a InvokeResponseIB within the matter protocol)
+///
+/// Invoke responses MUST contain exacltly ONE of:
+///   - response data (accessed via `ResponseEncoder`)
+///   - A status, which may be success or failure, both of which may
+///     contain a cluster-specific error code.
+///
+/// To encode a response, `Complete` MUST be called.
+///
+/// `Complete` requirements
+///   - Complete with InteractionModel::Status::Success will respond with data
+///     some response data was written.
+///   - Any other case (including success with cluster specific codes) implies
+///     no response data and a status will be encoded instead
+///       - this includes the case when some response data was written already.
+///         In that case, the response data will be rolled back and only the status
+///         will be encoded.
+///
+/// Creating a response MUST be retried at most once, if and only if `Complete`
+/// returns CHIP_ERROR_BUFFER_TOO_SMALL:
```

Comment:
nit, is it required to always retry? If not perhaps something like the following

```suggestion
/// Creating a response MAY be retried at most once, if and only if `Complete`
/// returns CHIP_ERROR_BUFFER_TOO_SMALL. Retry attempts MUST not exceed 1:
```